### PR TITLE
Improve spans in error messages around generic types

### LINKED ISSUE #1401: Improve spans for unresolved generics
In some cases, the spans we associate with the idents for generic types are at the declaration site, when it'd be better if they were at the call site. See the pictures in #1400 as an example. We'd want those errors to point at the calling of `foo()` in `fn main()` in that PR. 

This will most likely involve adding a `Span` to `Ref` types in the type engine. Currently, when a type `Ref`s to another type, the span we use is the type being referred to. The referrer loses its span info. Adding an (optional?) `Span` to `Ref` could give us this flexibility.

### PR DESCRIPTION
**Pull request by sezna**
This PR adds a `Span` to every `TypeInfo::Ref` to improve the error UX. Currently, if a type refers to another type via `Ref`, we use that underlying type as the span for errors and warnings. However, in the case of inference or monomorphized types, we want the call site to be shown. This PR uses spans from the call site preferentially and falls back to declaration spans if there is no `Ref` type preceding.

Tl;dr, before, the span pointed to the decl:
image

After:
image

This helps me a lot with #1628 and #1143, which is why I went ahead and knocked it out. 

Closes #1401

### CODE REVIEW

**Review by mohammadfawaz — APPROVED** [review]
I like it. Looks good!
